src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,276 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class OrderedRingBuffer(Generic[T]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Container,
+        resampling_period_s: float,
+        time_index_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: Instance of a buffer container to use internally.
+            resampling_period_s: Resampling period in seconds.
+            time_index_alignment: Arbitary point in time used to align
+                timestamped data with the index position in the buffer.
+                Used to make the data stored in the buffer align with the
+                beginning and end of the buffer borders.
+
+                For example, if the `time_index_alignment` is set to
+                "0001-01-01 12:00:00", and the `resampling_period_s` is set to
+                3600 (1 hour) and the length of the buffer is 24, then the data
+                stored in the buffer could correspond to the time range from
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date choosen
+                arbitrarily here).
+        """
+        self._buffer = buffer
+        self._resampling_period_s = resampling_period_s
+        self._time_index_alignment = time_index_alignment
+
+        self._missing_windows = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+
+    @property
+    def missing_windows(self) -> list:
+        """Get the list of missing windows.
+
+        The list of missing of windows consists of dicts with the members
+        "start" and "end" which are describing the time ranges of gaps, or in
+        other words, where we received data with `missing = True`.
+
+        Returns:
+            list of missing windows.
+        """
+        return self._missing_windows
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> bool:
```

Comment:
We could possibly use the Sample type defined [here](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/timeseries/_base_types.py)